zebra-state/src/service/check/difficulty.rs

```diff
@@ -0,0 +1,300 @@
+        if NetworkUpgrade::is_testnet_min_difficulty_block(
+            self.network,
+            self.candidate_height,
+            self.candidate_time,
+            self.relevant_times[0],
+        ) {
+            assert_eq!(
+                self.network,
+                Network::Testnet,
+                "invalid network: the minimum difficulty rule only applies on testnet"
+            );
+            ExpandedDifficulty::target_difficulty_limit(self.network).to_compact()
+        } else {
+            self.threshold_bits()
+        }
+    }
+
+    /// Calculate the `difficulty_threshold` for a candidate block, based on the
+    /// `candidate_height`, `network`, and the relevant `difficulty_threshold`s and
+    /// `time`s.
+    ///
+    /// See `expected_difficulty_threshold` for details.
+    ///
+    /// Implements `ThresholdBits` from the Zcash specification. (Which excludes the
+    /// Testnet minimum difficulty adjustment.)
+    fn threshold_bits(&self) -> CompactDifficulty {
+        let averaging_window_timespan = NetworkUpgrade::averaging_window_timespan_for_height(
+            self.network,
+            self.candidate_height,
+        );
+
+        let threshold = (self.mean_target_difficulty() / averaging_window_timespan.num_seconds())
+            * self.median_timespan_bounded().num_seconds();
+        let threshold = min(
+            ExpandedDifficulty::target_difficulty_limit(self.network),
+            threshold,
+        );
+
+        threshold.to_compact()
+    }
+
+    /// Calculate the arithmetic mean of the averaging window thresholds: the
+    /// expanded `difficulty_threshold`s from the previous `PoWAveragingWindow` (17)
+    /// blocks in the relevant chain.
+    ///
+    /// Implements `MeanTarget` from the Zcash specification.
+    fn mean_target_difficulty(&self) -> ExpandedDifficulty {
+        // In Zebra, contextual validation starts after Sapling activation, so we
+        // can assume that the relevant chain contains at least 17 blocks.
+        // Therefore, the `PoWLimit` case of `MeanTarget()` from the Zcash
+        // specification is unreachable.
+
+        let averaging_window_thresholds =
+            &self.relevant_difficulty_thresholds[0..POW_AVERAGING_WINDOW];
+
+        // Since the PoWLimits are `2^251 − 1` for Testnet, and `2^243 − 1` for
+        // Mainnet, the sum of 17 `ExpandedDifficulty` will be less than or equal
+        // to: `(2^251 − 1) * 17 = 2^255 + 2^251 - 17`. Therefore, the sum can
+        // not overflow a u256 value.
+        let total: ExpandedDifficulty = averaging_window_thresholds
+            .iter()
+            .map(|compact| {
+                compact
+                    .to_expanded()
+                    .expect("difficulty thresholds in previously verified blocks are valid")
+            })
+            .sum();
+        let total: U256 = total.into();
+        let divisor: U256 = POW_AVERAGING_WINDOW.into();
+
+        (total / divisor).into()
```

Comment:
are these `into` calls necessary? I know you did some fancy stuff with div impls that automatically call `Into`, just wanna double check that this code doesn't predate those changes.

Reply:
Yeah only the divisor `into` is needed, see commit 4eff86ef4681bf44d71a41ab8fdec9ebcdf8bc54